### PR TITLE
Added debounce for typeahead search optimization

### LINKED ISSUE #3472: Explore request timing optimizations for typeahead search and username/tag autocompletion
Our nice typeahead search is SUPER! (Thanks @milaaraujo @stefannibrasil !!)

I think we could speed it up a bit and also have it be a little less jittery by using a throttling strategy.

See how when you type the word "balloon" it makes 5 different queries? 

screen shot 2018-09-25 at 2 30 33 pm

We're using Bootstrap Typeahead: https://github.com/bassjobsen/Bootstrap-3-Typeahead to run this code. 

For starters, we can set the `minLength` and `delay` to 3 and 1, respectively, so it only starts searching after 3 characters are input, and only sends one request per second. This should slow things down a good bit (but test it out locally to see if it works for you!)

After merging those changes, we could explore more sophisticated "debounce" strategy -- https://davidwalsh.name/javascript-debounce-function

### PR DESCRIPTION
**Pull request by aashna27**
Fixes #3472

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts 📁
* [ ] PR is descriptively titled 📑
* [ ] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by aashna27** [comment]
@jywarren please have a look I have made the changes as per the issue from codeclimate https://github.com/AtomLinter/linter-eslint/issues/462 referring this link for issue problem but it still persists and I have added debounce as you suggested!!

File: app/assets/javascripts/debounce.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Returns a function, that, as long as it continues to be invoked, will not
+// be triggered. The function will be called after it stops being called for
+// N milliseconds. If `immediate` is passed, trigger the function on the
+// leading edge, instead of the trailing.
+export function debounce(func, wait, immediate) {
```
Comment (jywarren) [inline app/assets/javascripts/debounce.js]: Ah, i haven't used this syntax of import/export before... perhaps we should rely on concatenation here? Primitive, i know...

Comment (jywarren) [reply app/assets/javascripts/debounce.js]: So wait, we're building with sprockets, i believe... old fashioned -- is this export/import syntax something that should work with that? Traditionally what I'd do is just make a `debounce` function and then concatenate it in by adding a ref to `debounce.js` in `application.js` -- not very nice but it works.

Comment (jywarren) [reply app/assets/javascripts/debounce.js]: Oh i see, this is ES6? https://stackoverflow.com/questions/950087/how-do-i-include-a-javascript-file-in-another-javascript-file#950146

Does this work locally?

**Comment by aashna27** [comment]
@jywarren any suggestions?

**Comment by jywarren** [comment]
Hi, @aashna27! Hmm, did you see my comments here? https://github.com/publiclab/plots2/pull/4904#discussion_r262312253

Shall we give a try doing the more basic means of including rather than import/export? Thanks!

I think we could just define a function at the global scope instead of import/export. Not as nice but we can probably refactor this later. Thank you!

**Comment by jywarren** [comment]
OK, i think i got this running! I directly used the npm debounce library. I tested locally, it's very nice! Thanks so much @aashna27 for getting this started, it was very easy to build on your code.

**Comment by jywarren** [comment]
@icarito this is ready to be merged and deployed if it passes tests.

**Comment by jywarren** [comment]
extended it to username and tagname completion

**Comment by aashna27** [comment]
Happy to be of help. 😊 Thanks

**Pull request merged**